### DISCUSSION TITLE
nuget-package-for-mvc-pager/

### DISCUSSION THREAD

**Discussion by martijnboland** [martijnboland/blog#42] [Blog comments]
# nuget-package-for-mvc-pager/

https://blogs.taiga.nl/martijn/posts/nuget-package-for-mvc-pager/

**Comment by disqus-comments-migrator** [comment]
<em>Tom commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at mei 15 2011, 11:14</em>

---

<p>Hi Martijn,</p><p>I am having a problem with the pager when I define custom routes. For instance in your MvcPaging.Demo project I add in Global.asax two routes before the default route:</p><p>    routes.MapRoute("Test1", "Test1",<br>        new { controller = "Home", action = "About" },<br>        new[] { "MvcPaging.Demo.Controllers" });</p><p>    routes.MapRoute("Test2", "Test2",<br>        new { controller = "Paging", action = "Index" },<br>        new[] { "MvcPaging.Demo.Controllers" });</p><p>    // default route comes here</p><p>When I click then on "Simple Paging" on your start page I get redirected to "/Test2" which is fine because the action link points to controller "Paging" and action "Index".</p><p>However the page links do not point to "/Test2?page=n" but to "/Test1?page=n" (so I get redirected to the About page with my route configuration above when I click on a page link).</p><p>I have tracked down the issue to "GetVirtualPathForArea" in "GeneratePageLink()" which returns those wrong links (it seems to return simply the first entry in the routing table which is wrong one in this case).</p><p>Do you have an idea how to fix this?</p><p>Thanks for feedback and thanks for your great work!</p>

**Comment by disqus-comments-migrator** [comment]
<em>martijn commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at mei 16 2011, 02:30</em>

---

<p>I have no idea about the real problem, but adding the action name as route constraint to the custom routes does seem to fix things:</p><p>routes.MapRoute(“Test1″, “Test1″,<br>new { controller = “Home”, action = “About” },<br>new { action = "About" }<br>new[] { “MvcPaging.Demo.Controllers” });</p><p>Found this workaround at <a href="http://forums.asp.net/t/1630319.aspx/1?Wrong+URLs+retrieved+w+GetVirtualPathForArea+amp+GetVirtualPath" rel="nofollow noopener" title="http://forums.asp.net/t/1630319.aspx/1?Wrong+URLs+retrieved+w+GetVirtualPathForArea+amp+GetVirtualPath">http://forums.asp.net/t/1630319.aspx/1?Wrong+URLs+retrieved+w+GetVirtualPathForArea+amp+GetVirtualPath</a></p>

**Comment by disqus-comments-migrator** [comment]
<em>Tom commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at mei 16 2011, 03:03</em>

---

<p>OK, thanks, this helps. I've noticed that you also can add the controller as constraint instead of the action (which is even necessary if both actions have the same name but only a different controller):</p><p>routes.MapRoute("Test1", "Test1″,<br>    new { controller = "Home", action = "Index" },<br>    new { controller = "Home" }<br>    new[] { "MvcPaging.Demo.Controllers" });</p><p>I am wondering if one could use "ActionLink" instead of "GetVirtualPathForArea". I am also using this sorting helper here (<a href="http://www.codecapers.com/post/sorting-data-in-aspnet-mvc.aspx)" rel="nofollow noopener" title="http://www.codecapers.com/post/sorting-data-in-aspnet-mvc.aspx)">http://www.codecapers.com/post/sorting-data-in-aspnet-mvc.aspx)</a> and it delegates a part of the link generation to the ActionLink helper. This sorting helper doesn't have the described problem (works also with areas) and the links seem to be always correct, no matter how I define my routes. But my MVC knowledge isn't advanced enough to see if ActionLink could also be an option for the Paging helper or not.</p>

**Comment by disqus-comments-migrator** [comment]
<em>martijn commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at mei 16 2011, 03:20</em>

---

<p>If I remember well, I tried it first with ActionLink, but at the time (MVC pre-V1) there was some blocking issue, therefore I took the GetVirtualPath(ForArea) route.<br>It might very well be possible that ActionLink is the better option nowadays.</p>

**Comment by disqus-comments-migrator** [comment]
<em>shijuvar commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at juli 05 2011, 06:00</em>

---

<p>How to integrate the paging with styles and CSS?</p>

**Comment by disqus-comments-migrator** [comment]
<em>martijn commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at juli 05 2011, 07:27</em>

---

<p>@shijuvar: wrap the pager in a div with a specific class (for example '.pager') or id. Then you can style it easily:</p><p>.pager <br>{<br>}</p><p>.pager .disabled <br>{<br>}</p><p>.pager .current <br>{<br>}</p><p>.pager span, .pager a<br>{<br>}</p><p>etc...</p><p>The pager outputs 2 specific css classes: 'current' for the current page and 'disabled' for disabled hyperlinks.</p>

**Comment by disqus-comments-migrator** [comment]
<em>Paul commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at oktober 25 2011, 10:22</em>

---

<p>Hey I love your visual studio theme, what is it?<br>(also the pager is pretty cool)</p><p>Cheers</p><p>Paul</p>

**Reply by disqus-comments-migrator** [reply]
<em>martijn commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at oktober 26 2011, 12:42, in reply to Paul's [comment](https://github.com/martijnboland/blog/discussions/42#discussioncomment-8952888)</em>

---

<p>The visual studio theme is son of obsidian, <a href="http://studiostyl.es/schemes/son-of-obsidian" rel="nofollow noopener" title="http://studiostyl.es/schemes/son-of-obsidian">http://studiostyl.es/schemes/son-of-obsidian</a>.</p>

**Comment by disqus-comments-migrator** [comment]
<em>Jan Bakker commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at december 03 2011, 10:43</em>

---

<p>Hello Martijn:</p><p>I am not sure this pager addresses the real problem. Say, you have 5 million records. Your search returns 2 million. Now you wish to display the first 20 records and then the next 20 and so on.</p><p>The problem I see is that it retrieves the full 2 million records and then 'takes' 20 and throws away the rest.</p><p>Your thoughts are appreciated.</p><p>Jan</p>

**Comment by disqus-comments-migrator** [comment]
<em>martijn commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at december 03 2011, 04:39</em>

---

<p>@Jan<br>The Pager Helper itself doesn't do anything with data. You just tell it how large your pagesize is, which page we're on and the total number of items.</p><p>The PagedList&lt;t&gt; that comes with the Pager can be created from an IQueryable&lt;t&gt; (returned from NHibernate or EF) which means that actual paging is executed via the database query.<br>Alternatively, you could even use a PagedList, set its properties and perform all the paging logic yourself.</p>
